### PR TITLE
feat: remirror for note editing

### LINKED ISSUE #769: Feature: Use Remirror for notecontent box
### Is your feature request related to a problem? Please describe.

The [note form](https://github.com/open-sauced/open-sauced/blob/master/src/components/NoteForm.js#L58-L90) could use an update with the Social mirror editor.

### Describe the solution you'd like

Adding the social editor would provide the ability to create emojis/ 

```js
import React from 'react';
import { SocialEmojiComponent, SocialProvider } from 'remirror/react/social';

const Editor = () => {
  return <SocialProvider><SocialEmojiComponent></SocialProvider>
}
```

https://github.com/remirror/remirror

### Additional context

example social remirror

cc/ @ifiokjr is this social editor ready for prime time?

### PR DESCRIPTION
**Pull request by mtfoley**
## What type of PR is this? (check all applicable)

- [ ] ♻️ Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [x] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

Congrats to @ifiokjr on release of v1.0.0 of remirror! 
I wanted to get started on a PR draft for the remirror note editing feature, since @0-vortex helped tremendously in cleaning up npm dependencies to free up this feature and others.

Todo:
- [x] Support and render some subset of markdown in editor
- [x] Manage state and initial content correctly, I'm still trying to get my head around the docs on this [https://remirror.io/docs/react/controlled](https://remirror.io/docs/react/controlled)
- [x] Implement social editor features. 
- [ ] ~~Add to storybook~~ NoteForm isn't currently in StoryBook (makes sense for open form control?)
- [x] Add/update docs
- [ ] ~~Add test similar to what's in [https://github.com/remirror/remirror/blob/main/packages/remirror__extension-markdown/__tests__/markdown-extension.spec.ts](https://github.com/remirror/remirror/blob/main/packages/remirror__extension-markdown/__tests__/markdown-extension.spec.ts). I'm skipping the existing NoteForm test on this first commit.~~ Sticking with "no violations on render" test?

This PR closes #769

## Mobile & Desktop Screenshots/Recordings
Screen Shot 2021-07-20 at 10 50 56 PM

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 docs
- [x] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?
turtle going up a slide

### CODE REVIEW

**Comment by mtfoley** [comment]
@bdougie @0-vortex @filiptronicek looks like Netlify deploy is choking on my CSS import in `src/components/NoteFormSocial.js`. Works ok for Webpack when running `npm start` locally, so I don't know what I'll need to do to make this work. I'd be referring to the remirror docs to get this working: [https://remirror.io/docs/installation#styling-the-editor](https://remirror.io/docs/installation#styling-the-editor)

**Comment by 0-vortex** [comment]
I'll have a swing at it in the morning 🤤

**Comment by 0-vortex** [comment]
> Installing remirror gives you access to all the extensions and core packages that are actively developed in the remirror repository.

Installing packages from both sources makes it very difficult to figure out exactly what the tree should look like - suggest installing just the required packages if you want a readable tree :D

**Comment by ifiokjr** [comment]
@mtfoley I'll be able to give you a hand on this in a few hours time 😊

**Comment by mtfoley** [comment]
Three steps forward, two steps back - deploy preview works on Netlify now. I added the `MentionExtension` and the `EmojiExtension`.  The MentionExtension doesn't seem to be doing anything, but I'm not sure if I need to feed it a set of valid names for mentioning. The EmojiExtension makes the app crash when running locally and I trigger it with by typing the activation character `:`.

@ifiokjr any pointers you'd like to give, would love it!

**Comment by mtfoley** [comment]
FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
Screen Shot 2021-08-01 at 10 08 28 PM

**Comment by 0-vortex** [comment]
> FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
> Screen Shot 2021-08-01 at 10 08 28 PM

gonna pick this up after work and let you know !

**Comment by mtfoley** [comment]
I hit up the remirror discord so we will see if anything comes back.

**Comment by 0-vortex** [comment]
> FWIW, here's the error that the emojis extension throws whenever I start a word with ":"
> Screen Shot 2021-08-01 at 10 08 28 PM

this is because of https://github.com/remirror/remirror/blob/5764fde3bc4643cb8c6de57ce4f3d2baa3654361/packages/remirror__extension-emoji/src/emoji-extension.ts#L37-L47 the emoji selection character is the end selection handler for remirror - configuring this plugin to use a different shortcut will remove this error but I have not produced a full working build yet

**Comment by mtfoley** [comment]
Got this back on the remirror discord:
> Could you try this?

```js
import data from 'svgmoji/emoji.json';

new EmojiExtension({ data, moji: 'noto' })
```

> I believe the README of `extension-emoji` has some mistakes. Will fix.

**Comment by mtfoley** [comment]
Ok I kind of got things to work with a snippet like that. However the suggestion box doesn’t come up but it will render an emoji you type. Trouble is, it ends up being added as an svg image rather than as the emoji itself so the markdown has these huge images interspersed with the text. The other side effect of this is that when you open the editor a second time to edit the notes the emojis are effectively lost because it seems I don’t have the editor ready to support those. I’ll push the changes up later today so y’all can see what im talking about.

**Comment by mtfoley** [comment]
Emojis work much better with plainText:true option. Still not seeing suggestions popup yet.

Other kink to mess with is whether editor will work correctly in dark mode. if you type ## then hit space it starts using heading 2 but in black text on dark background. Need to see if there's some step required to make markdown extension aware of colors in use... maybe?

**Comment by mtfoley** [comment]
Got emoji suggestion popup working, and changed gears on styling method in the process (using `AllStyledComponent` instead of CSS import).  I found this closed issue in the remirror repo and based on the fact that the person wasn't told they needed to make any changes, I assumed they were pretty close: https://github.com/remirror/remirror/issues/968

Emoji suggestion popup has a transparent background which I'm not sure is right. It's also staying inside the editor's outline, so I'm not sure that's right either. 
Screen Shot 2021-08-07 at 3 16 00 PM


I did find that remirror styles headings with a CSS color variable, so I may be able to use some CSS rules to account for this.

I think another change that I'm going to need is to pull the "Edit Notes" and "Cancel" buttons inside the Remirror component so that I can mess with state more easily 🤞 

This thing is turning into spaghetti, so I'll probably break it up into a component once something works well.

**Comment by mtfoley** [comment]
Progress so far:
- added CSS rules to handle menu background color and making editor bigger so more of the menu would fit.  
- preloading the SVG image from with `display:none` so that the emoji menu is snappier when it's first used. 
- added the code extension too in order to allow inline code blocks. I don't think it does multiline or language specific blocks.

**Comment by mtfoley** [comment]
@0-vortex I applied those fixes you put up the other day, and merged latest upstream 🙏

If anyone can take a look at my state mgmt I'd appreciate it.  When you click cancel while editing, the notes go blank.  Apparently I'm not setting "previouslySavedValue" correctly.

**Comment by 0-vortex** [comment]
> @0-vortex I applied those fixes you put up the other day, and merged latest upstream 🙏
> 
> If anyone can take a look at my state mgmt I'd appreciate it. When you click cancel while editing, the notes go blank. Apparently I'm not setting "previouslySavedValue" correctly.

Can take a look after work, not thrilled about state management so maybe link all the issues and we finish it today? Maybe get @bdougie to look at this on stream?

**Comment by mtfoley** [comment]
@0-vortex Finally peer pressured me into taking this thing out of draft status. Last bit of context:
- I was trying out tests last night. Two snags for replicating existing tests, and I think we'd need to add another dependency to replicate test as-is and I'm not sure how it would actually work. At the moment it passes the "no violations" test.
Still need to add to storybook.

**Comment by 0-vortex** [comment]
### Tooling wise, how to make your PR the best PR possible 🔥 

#### Step 1: restore the proper lockfiles

```shell
git restore --source main -- npm-shrinkwrap.json package.json
```


#### Step 2: remirror perfect life

Then, remember all the support you got from `@remirror` discord and start to verify their `package.json` and go full minimalist, installing only the absolutely required packages.

```shell
npm i -S @remirror/react @remirror/styles @remirror/extension-bold @remirror/extension-code @remirror/extension-heading @remirror/extension-italic @remirror/extension-list @remirror/extension-markdown @remirror/extension-tables
```


#### Step 3: remove the `remirror/extensions` import

```javascript
import {CodeExtension} from "@remirror/extension-code";
import {BulletListExtension} from "@remirror/extension-list";
import {HeadingExtension} from "@remirror/extension-heading";
import {BoldExtension} from "@remirror/extension-bold";
import {ItalicExtension} from "@remirror/extension-italic";
import {TableExtension} from "@remirror/extension-react-tables";
import {MarkdownExtension} from "@remirror/extension-markdown";
import {EmojiExtension} from "@remirror/extension-emoji";
```

**Comment by 0-vortex** [comment]
LGTM!

**Comment by mtfoley** [comment]
in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛  Pushed changes and removed old files.

Current form isn't in Storybook. Keeping no violations test pending review.

**Comment by 0-vortex** [comment]
> in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> 
> Current form isn't in Storybook. Keeping no violations test pending review.

What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱

**Comment by mtfoley** [comment]
> > in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> > Current form isn't in Storybook. Keeping no violations test pending review.
> 
> What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱

Says it in the docs [https://git-scm.com/docs/git-restore](https://git-scm.com/docs/git-restore)... which I only read bc my version of git apparently wasn't new enough so I ended up doing `wget -O package.json https://raw.githubusercontent.com/open-sauced/open-sauced/main/package.json` 🥇

**Comment by mtfoley** [comment]
BTW, still gotta update README.md at least...

**Comment by 0-vortex** [comment]
> > > in such shock and awe that @0-vortex had me try out an experimental `git restore` command 😛 Pushed changes and removed old files.
> > > Current form isn't in Storybook. Keeping no violations test pending review.
> > 
> > 
> > What experimental? Standard procedure for solving pull request issues quickly 🔥📖🪱
> 
> Says it in the docs https://git-scm.com/docs/git-restore... which I only read bc my version of git apparently wasn't new enough so I ended up doing `wget -O package.json https://raw.githubusercontent.com/open-sauced/open-sauced/main/package.json` 🥇

Oh snap I apparently don't have experimental in my vocabulary, that's a good first-time issue for updating the pull request docs. Or you could do it since you discovered it! 👯

File: src/components/NoteFormSocial.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  useEmoji();
+  return (
+    <EditorComponent />
+  );
+};
+const Menu = () => {
+  // Access the commands and the activity status of the editor.
+  const commands = useCommands();
+  const startNewNote = () => {
+    commands.insertMarkdown("### Notes for Repo\n- fork the repo\n\n- follow everybody on twitter\n-troll, troll, troll<Cursor>");
```
Comment (mtfoley) [inline src/components/NoteFormSocial.js:47]: @bdougie this is a bit of an easter egg. anyway, I'm undecided about whether to keep this button in place (@0-vortex had suggested changing the label on it), as opposed to just pre-loading some template markdown when someone starts editing a blank note.

Comment (0-vortex) [reply src/components/NoteFormSocial.js:47]: freedom!

File: src/index.css [hunk 2]
```diff
@@ -117,4 +117,20 @@ body.dark .react-loading-skeleton{
     var(--backgroundGrey)
   );
 }
-
+.remirror-editor {
+  min-height: 250px !important;
```
Comment (bdougie) [inline src/index.css:121]: Why the important here. Is this because remirror does not play nicE?

**Review by bdougie — COMMENTED** [review]
Just one question then I can merge

File: src/index.css (hunk 2 above)
Comment (mtfoley) [reply src/index.css:121]: Editor stayed small and emoji list stayed too small to be usable. you would only see like 3 options

File: src/index.css [hunk 3]
```diff
@@ -117,4 +117,20 @@ body.dark .react-loading-skeleton{
     var(--backgroundGrey)
   );
 }
-
+.remirror-editor {
```
Comment (bdougie) [inline src/index.css:120]:
```suggestion
// important is to manage the emoji sizing, so more appear while searching. 
.remirror-editor {
```

File: src/components/NoteFormSocial.js (hunk 1 above)
Comment (bdougie) [reply src/components/NoteFormSocial.js:47]: Let's name the button, "basic template"

Comment (mtfoley) [reply src/components/NoteFormSocial.js:47]: 👍

**Review by bdougie — APPROVED** [review]
Looking good, we will need to open issues on the UI and CSS.

**Pull request merged**